site/content/blog/kapp-versioned-resource.md

```diff
@@ -0,0 +1,566 @@
+
+Let's play with some example:
+```yaml 
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: crd-config
+  annotations:
+    kapp.k14s.io/versioned: ""
+data:
+  hello_msg: hello-kapp
+
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: simplecrds.example.com
+spec:
+  group: example.com
+  versions:
+    - name: v1beta1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                name:
+                  type: string
+                image:
+                  type: string
+  scope: Namespaced
+  names:
+    kind: simplecrd
+    plural: simplecrds	
+
+---
+apiVersion: "example.com/v1beta1"
+kind: simplecrd
+metadata:
+  name: first-cr
+spec:
+  name: simple-app
+  image: docker.io/dkalinin/k8s-simple-app:latest
+```
+
+Let's deploy them using `kapp`:
+```bash
+$ kapp deploy -a crd-app -f version-crd.yaml                           
+Target cluster 'https://127.0.0.1:33907' (nodes: minikube)
+
+Changes
+
+Namespace  Name                    Kind                      Age  Op      Op st.  Wait to    Rs  Ri  
+(cluster)  simplecrds.example.com  CustomResourceDefinition  -    create  -       reconcile  -   -  
+default    crd-config-ver-1        ConfigMap                 -    create  -       reconcile  -   -  
+^          first-cr                simplecrd                 -    create  -       reconcile  -   -  
+
+Op:      3 create, 0 delete, 0 update, 0 noop, 0 exists
+Wait to: 3 reconcile, 0 delete, 0 noop
+
+Continue? [yN]: y
+```
+Let's update the value of `data.hello_msg` to `hello-carvel` in configmap `crd-config` and redeploy the app `crd-app`.
+
+```bash
+$ kapp deploy -a crd-app -f version-crd.yaml --diff-changes
+Target cluster 'https://127.0.0.1:33907' (nodes: minikube)
+
+@@ create configmap/crd-config-ver-2 (v1) namespace: default @@
+  ...
+  1,  1   data:
+  2     -   hello_msg: hello-kapp
+      2 +   hello_msg: hello-carvel
+  3,  3   kind: ConfigMap
+  4,  4   metadata:
+
+Changes
+
+Namespace  Name              Kind       Age  Op      Op st.  Wait to    Rs  Ri  
+default    crd-config-ver-2  ConfigMap  -    create  -       reconcile  -   -  
+
+Op:      1 create, 0 delete, 0 update, 0 noop, 0 exists
+Wait to: 1 reconcile, 0 delete, 0 noop
+
+Continue? [yN]:y
+```
+
+
+Now what we want is whenever there is any change happen in configmap `crd-config` the custom resource `first-cr` should get updated or restarted. To achive this we have to add annotation `kapp.k14s.io/versioned-explicit-ref` to the custom resource so that on every update in configmap `crd-config` kapp will make update to custom resource as well.
```

Comment:
```suggestion
Now what we want is whenever there is any change happen in ConfigMap `crd-config` the custom resource `first-cr` should get updated or restarted. To achieve this we have to add annotation `kapp.k14s.io/versioned-explicit-ref` to the custom resource so that on every update in ConfigMap `crd-config` kapp will make update to custom resource as well.
```